pkg/compose/convergence.go

```diff
@@ -328,7 +330,33 @@ func mustRecreate(expected types.ServiceConfig, actual moby.Container, policy st
 	}
 	configChanged := actual.Labels[api.ConfigHashLabel] != configHash
 	imageUpdated := actual.Labels[api.ImageDigestLabel] != expected.CustomLabels[api.ImageDigestLabel]
-	return configChanged || imageUpdated, nil
+	if configChanged || imageUpdated {
+		return true, nil
+	}
+
+	if c.networks != nil {
+		// check the networks container is connected to are the expected ones
+		for net := range expected.Networks {
+			id := c.networks[net]
+			if id == "swarm" {
+				// corner-case : swarm overlay network isn't visible until a container is attached
+				continue
+			}
+			found := false
+			for _, settings := range actual.NetworkSettings.Networks {
+				if settings.NetworkID == id {
+					found = true
+					break
+				}
+			}
+			if !found {
+				// config is up-t-date but container is not connected to network - maybe recreated ?
```

Comment:
Fwiw this makes a lot of sense to me. Would even say it makes sense to recreate if it is connected to undesired networks.